AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLDocument/SyncEnabledGraphQLDocument.swift

```diff
@@ -0,0 +1,82 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+import Foundation
+
+public class SyncEnabledGraphQLDocument<Document: GraphQLDocument>: GraphQLDocument {
+
+    private let graphQLDocument: Document
+    private let version: Int?
+
+    public init(_ graphqQLDocument: Document, version: Int? = nil) {
+        self.graphQLDocument = graphqQLDocument
+        self.version = version
+    }
+
+    public var documentType: GraphQLDocumentType {
+        graphQLDocument.documentType
+    }
+
+    public var name: String {
+        graphQLDocument.name
+    }
+
+    public var modelType: Model.Type {
+        graphQLDocument.modelType
+    }
+
+    public var stringValue: String {
+        let selectionSetString = selectionSetFields.map { $0.toString() }.joined(separator: "\n    ")
+        if let inputTypes = inputTypes, let inputParameters = inputParameters {
+            return """
+            \(documentType) \(name.pascalCased())(\(inputTypes)) {
+              \(name)(\(inputParameters)) {
+                \(selectionSetString)
+              }
+            }
+            """
+        }
+
+        return """
+        \(documentType) \(name.pascalCased()) {
+          \(name) {
+            \(selectionSetString)
+          }
+        }
+        """
+    }
+
+    public var inputTypes: String? {
+        graphQLDocument.inputTypes
+    }
+
+    public var inputParameters: String? {
+        graphQLDocument.inputParameters
+    }
+
+    public var variables: [String: Any] {
+        var variables = graphQLDocument.variables
+
+        if let version = version, var input = variables["input"] as? [String: Any] {
+            input.updateValue(version, forKey: "_version")
+            variables["input"] = input
+        }
+
+        return variables
+    }
+
+    public var selectionSetFields: [SelectionSetField] {
+        if graphQLDocument.name.contains("list") {
```

Comment:
Relying on string comparison in document names is a bug waiting to happen. We need a more robust and explicit way of determining that a selection set field returns a `List`, and specifically one of the AWS-defined `List`s that have an `items` and `nextToken` field for pagination.

Reply:
ah yes, the problem here is if name contains "list" where it was part of the user generated model's name then this all breaks. taking a look

Reply:
hmm so i basically moved the responsibility of knowing whether it is a paginated selection set to the `[SelectionSet]` extension like `[SelectionSet].paginate()` will return append "items" and "nextToken" and `[SelectionSet].isPaginated() -> Bool` checks if first and last items in the array. may have to chat with you about where you see the custom `List` used